### PR TITLE
[spirv] Make default configuration consider all partitioned loops

### PR DESCRIPTION
**Pull request by antiagainst**
This commit revises the default linalg op configuration to consider
all partitioned loops when deciding workgroup size and tiling
schemes. Previously we only consider the innermost partitioned loop
and distribute that to 1-D workgroup.

This helps to fill the GPU for the cases where we have small innermost
dimensions, like 244x224x3. Under the previous configuration, we will
just distribute to a workgorup of, like, 64x1x1, which starves the
GPU and pushes all the work to effectively 3 threads. With the current
configuration this would be distributed to a workgroup of 1x32x2.

### CODE REVIEW

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 1]
```diff
@@ -319,24 +320,87 @@ static LogicalResult setOpConfig(spirv::ResourceLimitsAttr limits,
...
+  // surrounding the op should match. Otherwise, something is incorrect.
+  assert(partitionedLoops.size() == tiledLoopInfo.size());
+
+  // The upper bound for each implicit loop: 0 - untiled, negative - dynamic.
+  SmallVector<int64_t> loopBounds(loopDepth, 0);
+  // tiledLoopInfo uses the reverse order of partitionedLoops.
+  for (auto pair : llvm::zip(llvm::reverse(partitionedLoops), tiledLoopInfo)) {
+    unsigned loopIndex = std::get<0>(pair);
+    const TiledLoopInfo &loopInfo = std::get<1>(pair);
+    if (auto attr = loopInfo.ub.dyn_cast<Attribute>()) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/SPIRV/KernelConfig.cpp:398]: Could replace this with some utility methods from here https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h

**Review by MaheshRavishankar — APPROVED** [review]
Few nits, but otherwise looks good to me.

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/KernelConfig.cpp]: Done

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 2]
```diff
@@ -319,24 +320,87 @@ static LogicalResult setOpConfig(spirv::ResourceLimitsAttr limits,
...
+      workgroupTileSizes[loopIndex] = threadTileSizes[loopIndex] = 1;
+    }
+    // Override the innermost dimension to distribute to threads in a subgroup.
+    workgroupTileSizes.back() = subgroupSize;
+    threadTileSizes.back() = 1;
+  };
+
+  // Special case for non-linalg ops.
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp || linalgOp.getNumOutputs() != 1) {
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/KernelConfig.cpp:368]: In the following when deducing whether the linalg op is vectorizable we only look at the first output. This is actually just to place it safely. Can certainly relax once we have more multi-output linalg ops.

[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/SPIRV/KernelConfig.cpp:368]: Oh, sorry. My brain interpreted it as handling the case of 0 outputs. This makes sense.

**Pull request merged**